File: backend/internal/domain/reasoning.go
package domain

type ReasoningMethod string

const (
	ReasoningMethodDirect      ReasoningMethod = "direct"
	ReasoningMethodStepByStep  ReasoningMethod = "step_by_step"
	ReasoningMethodMetaPrompt  ReasoningMethod = "meta_prompt"
	ReasoningMethodExpertPanel ReasoningMethod = "expert_panel"
)

func (m ReasoningMethod) IsValid() bool {
	return m == ReasoningMethodDirect ||
		m == ReasoningMethodStepByStep ||
		m == ReasoningMethodMetaPrompt ||
		m == ReasoningMethodExpertPanel
}

type ModelRequest struct {
	Model        string
	SystemPrompt string
	UserPrompt   string
	Messages     []ModelMessage
	JSON         bool
	MaxTokens    int
	Temperature  *float64
}

type ModelMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type ModelResponse struct {
	Content      string
	Model        string
	FinishReason string
	Usage        Usage
}

type ExpertSolution struct {
	Role   string `json:"role"`
	Answer string `json:"answer"`
}

type ReasoningAttempt struct {
	Method          ReasoningMethod  `json:"method"`
	Answer          string           `json:"answer"`
	GeneratedPrompt string           `json:"generatedPrompt,omitempty"`
	Experts         []ExpertSolution `json:"experts"`
	Model           string           `json:"model"`
	FinishReason    string           `json:"finishReason"`
	Usage           Usage            `json:"usage"`
}

type ReasoningScore struct {
	Method       ReasoningMethod `json:"method"`
	Correctness  int             `json:"correctness"`
	Clarity      int             `json:"clarity"`
	Verification int             `json:"verification"`
	Feedback     string          `json:"feedback"`
}

type ReasoningReview struct {
	Winner          ReasoningMethod  `json:"winner"`
	Verdict         string           `json:"verdict"`
	ReferenceAnswer string           `json:"referenceAnswer"`
	Differences     []string         `json:"differences"`
	Scores          []ReasoningScore `json:"scores"`
	Model           string           `json:"model"`
	Usage           Usage            `json:"usage"`
}

File: ios/AIChallenge/Domain/Models/Reasoning.swift
import Foundation

enum ReasoningMethod: String, Codable, CaseIterable, Identifiable, Sendable {
    case direct
    case stepByStep = "step_by_step"
    case metaPrompt = "meta_prompt"
    case expertPanel = "expert_panel"

    var id: Self { self }
}

struct ReasoningUsage: Codable, Equatable, Sendable {
    let promptTokens: Int
    let completionTokens: Int
    let totalTokens: Int
}

struct ExpertSolution: Codable, Equatable, Sendable {
    let role: String
    let answer: String
}

struct ReasoningAttempt: Codable, Equatable, Sendable {
    let method: ReasoningMethod
    let answer: String
    let generatedPrompt: String?
    let experts: [ExpertSolution]
    let model: String
    let finishReason: String
    let usage: ReasoningUsage

    init(
        method: ReasoningMethod,
        answer: String,
        generatedPrompt: String? = nil,
        experts: [ExpertSolution] = [],
        model: String,
        finishReason: String,
        usage: ReasoningUsage
    ) {
        self.method = method
        self.answer = answer
        self.generatedPrompt = generatedPrompt
        self.experts = experts
        self.model = model
        self.finishReason = finishReason
        self.usage = usage
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        method = try container.decode(ReasoningMethod.self, forKey: .method)
        answer = try container.decode(String.self, forKey: .answer)
        generatedPrompt = try container.decodeIfPresent(String.self, forKey: .generatedPrompt)
        experts = try container.decodeIfPresent([ExpertSolution].self, forKey: .experts) ?? []
        model = try container.decode(String.self, forKey: .model)
        finishReason = try container.decode(String.self, forKey: .finishReason)
        usage = try container.decode(ReasoningUsage.self, forKey: .usage)
    }
}

struct ReasoningScore: Codable, Equatable, Sendable {
    let method: ReasoningMethod
    let correctness: Int
    let clarity: Int
    let verification: Int
    let feedback: String

    var average: Double {
        Double(correctness + clarity + verification) / 3
    }
}

struct ReasoningReview: Codable, Equatable, Sendable {
    let winner: ReasoningMethod
    let verdict: String
    let referenceAnswer: String
    let differences: [String]
    let scores: [ReasoningScore]
    let model: String
    let usage: ReasoningUsage
}

struct ReasoningComparison: Equatable, Sendable {
    let problem: String
    let attempts: [ReasoningAttempt]
    let review: ReasoningReview

    func attempt(for method: ReasoningMethod) -> ReasoningAttempt? {
        attempts.first { $0.method == method }
    }
}
